src/main/scala/com/astrolabsoftware/grafink/common/PartitionManager.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 AstroLab Software
+ * Author: Yash Datta
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.astrolabsoftware.grafink.common
+
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+import scala.annotation.tailrec
+
+import org.apache.hadoop.fs.{ FileSystem, Path }
+import org.apache.spark.sql.SparkSession
+import zio.{ RIO, Task, ZIO }
+import zio.logging.{ log, Logging }
+
+case class PartitionPath(year: String, month: String, day: String)
+
+case class PartitionManager(startDate: LocalDate, duration: Int) {
+
+  import PartitionManager._
+
+  implicit def toPartitionPath(date: LocalDate): PartitionPath =
+    PartitionPath(
+      year = s"${f"${date.getYear}%02d"}",
+      month = paddedInt(date.getMonth.getValue),
+      day = paddedInt(date.getDayOfMonth)
+    )
+
+  @tailrec
+  private def paths(startDate: LocalDate, duration: Int, p: List[PartitionPath]): List[PartitionPath] = {
+    if (duration <= 0) return p
+    val nextDate = startDate.plus(1, ChronoUnit.DAYS)
+    paths(nextDate, duration - 1, nextDate :: p)
+  }
+
+  val partitionPaths: List[PartitionPath] = paths(startDate, duration - 1, startDate :: Nil)
+
+  def testDirExist(fs: FileSystem, path: String): ZIO[Logging, Throwable, Boolean] = {
+    val p = new Path(path)
+    for {
+      result <- ZIO.effect(fs.exists(p) && fs.getFileStatus(p).isDirectory)
+      // TODO: Maybe some better way of printing this log?
+      _ <- if (result == false) log.warn(s"Filtering out $path as no such directory exists") else ZIO.succeed(Unit)
+    } yield result
+  }
+
+  def getValidPartitionPathStrings(basePath: String, fs: FileSystem): ZIO[Logging, Throwable, List[String]] = {
+    val validPaths       = partitionPaths
+    val validPathStrings = validPaths.map(toPathString(basePath, _))
+    ZIO.filter(validPathStrings)(testDirExist(fs, _))
+  }
+
+  def deletePartitions(basePath: String, fs: FileSystem): ZIO[Logging, Throwable, Unit] =
+    for {
+      paths <- getValidPartitionPathStrings(basePath, fs)
+    } yield ZIO
+      .collectAll(paths.map { path =>
+        RIO
+          .effect(fs.delete(new Path(path), true))
```

Comment:
This function will be used only when overwriting data, we cannot use spark overwrite mode because in that case, it deletes all the data (as far as I know) , whereas what we might want is 'selective' overwrite.

Reply:
But which data we want to overwrite, the grafink generated data or original fink data?

Reply:
grafink generated data, in case we re-run for the same day for some reason

Reply:
OK!